### ISSUE TITLE
MacOS will now default to ZSH

### ISSUE THREAD

**Issue by Carreau** [swcarpentry/shell-novice#978]
Hello there; 

It has been announces that the next MacOS will have Zsh as the [default shell](https://support.apple.com/en-us/HT208050) this lesson assume that by default the shell on macos is bash. 

It should not matter much for most command, but might introduce some edge cases in for loops etc...

The maintainer should likely decide whether it is worth covering some of Zsh, or setup macos to use Bash

**Comment by gcapes** [comment]
Thanks for sharing this information!

**Comment by gdevenyi** [comment]
Will need to look into if bash will remain installed, if so, we can mitigate.

More OSX/Windows/Linux differences :*(

**Comment by gcapes** [comment]
It seems that it will remain available for the next release at least: https://www.techtelegraph.co.uk/apple-replaces-bash-with-zsh-as-the-default-shell-in-macos-catalina/

**Comment by LHurst-UoB** [comment]
zsh is bash-compatible for most things - it may be worth checking what (if any) exercises won't work with zsh.

Personally I much prefer zsh, it's much more user-friendly and forgiving than bash especially with plug-ins (that are usually included out of the box) for typo-correction suggestions and tab-completion of arguments for commands that provide a parseable '--help' output.

Another reflection I had was that the lesson is "The Unix Shell", not "BASH".  There are many Unix shells, and many of them are sufficiently BASH-like that the lesson works fine with them.

**Comment by trickytank** [comment]
My personal opinion is that they'd be better off just using the default zsh if the exercises still work. 
I tried some exercises in **Shell Scripts** and **Loops**.

Errors may be handled by zsh differently. 
This affects some examples which are errors, but also for mistakes students are likely to make.
Examples that may happen in lessons include:

- Directories may be changed without `cd`. e.g. `cd Desktop` can be `Desktop`
- Wildcards that don't match any files are caught by zsh, where as bash just gives the input strings to the command. This includes _for loops_, such that it won't try to run a loop with the wildcard. 
This means the `cp *.dat original-*.dat` example in Loops has the wrong explanation in zsh, as zsh causes the error. 

zsh doesn't require quotes around variables to keep them as a single argument. 
This gives a different result in one of the Loops examples:

**bash:**
```
$ for filename in "red dragon.dat" "purple unicorn.dat"
> do
>     head -n 100 $filename | tail -n 20
> done
```
head: red: No such file or directory
head: dragon.dat: No such file or directory
head: purple: No such file or directory
CGGTACCGAA
AAGGGTCGCG
CAAGTGTTCC
CGGGACAATA
GTTCTGCTAA
GATAAGTATG
TGCCGACTTA
CCCGACCGTC
TAGGTTATAA
GGCACAACCG
CTTCACTGTA
GAGGTGTACA
AGGATCCGTT
GCGCGGGCGG
CAGTCTATGT
TTTTCGACAC
TGGACTGCTT
CCCTTTGAGG
GTGGATTTTT
CGTAACGGGT

**zsh:**
```
$ for filename in "red dragon.dat" "purple unicorn.dat"
do
    head -n 100 $filename | tail -n 20
done

head: red dragon.dat: No such file or directory
head: purple unicorn.dat: No such file or directory
```

**Comment by gcapes** [comment]
Thanks for testing these examples.

**Comment by edbennett** [comment]
Since macOS Catalina is now public and being shipped with new machines, it's becoming increasingly likely that learners will encounter this, e.g. if they have bought a new Mac at the start of their PhD. Because of this, it'd be good if this can be resolved soonish.

Something I've observed on my machine is that if you upgrade from a previous version of macOS, then it sticks with the default terminal from that installation (usually Bash), and prints a message indicating that zsh is now the default when you open a terminal. (I've not tested if this happens if you haven't ever opened Terminal before upgrading.)

It's also worth recognising that the version of bash in macOS is very out of date, so is already a little inconsistent with the instructions (e.g. `ls --help` fails, which isn't a great look for one of the first commands that learners may type)

Options I can think of:
• Get Mac users to set their terminal to bash for the workshop. (Not ideal, since we're dragging them away from a better shell experience just so they can follow the bash lesson material)
• Have a toggle switch for which terminal the learner is using, similar to what HPC Carpentries does for queue manager. (Not ideal, since it complicates instruction significantly, and imposes extra cognitive load on learners to know what shell they have)
• Get Linux and Windows users to install zsh and adjust the lesson material to teach zsh instead of bash (not ideal, since bash is the default shell for most systems so at some point users will need to interact with it, e.g. if they use an HPC system that doesn't have zsh installed)
• Rewrite the examples so that they fail the same way in bash and zsh (not ideal, since this will then fail to teach bash users the most common failure modes and how to avoid them)
• Add notes after the examples that work differently in zsh explaining that zsh (or macOS Catalina) is very helpful and stops this breaking, but bear it in mind for running on other machines (not ideal, since we're teaching users to run shell scripts as an argument to `bash`, so scripts may fail when running the commands interactively doesn't)

No solution I can think of is ideal. Does anyone else have other ideas I've overlooked? Or opinions on which is the least of all evils?

**Comment by LHurst-UoB** [comment]
> It's also worth recognising that the version of bash in macOS is very out of date, so is already a little inconsistent with the instructions (e.g. ls --help fails, which isn't a great look for one of the first commands that learners may type)

This has nothing to do with bash and is a BSD (upon which Darwin/OSX is built) vs GNU user-space (which is the most common user-space used with the Linux kernel) difference, `ls` is not a bash command, it's a separate (and different, on each) program on both systems.

For most exercises they success/fail similarly on both shells - as the error messages are sometimes inconsistent across platforms/versions even within bash I'm not sure it's worth getting so hung-up on "bash words the error 'thus' but zsh says 'that'" aspect unless there's something that is actually going to confuse attendees.

We should be teaching people to *read* the error messages they receive anyway, so I don't think that it being subtly different from the exact wording in the material (again, which we already encounter with Bash anyway) is a problem.

IMO we have an issue is if some examples do not work at all (or vice-versa if supposed to fail) on zsh and I'm yet to see a single example of that (which we can then collaborate on fixing).

**Comment by LHurst-UoB** [comment]
(having just setup for a new workshop): the setup directions already tell Linux user's to run Bash (open terminal, type 'bash') if Bash is not their default shell.

Why not just replicate this for macOS for now (which currently still says "bash is the default for all versions of macOS), while we figure out the longer-term solution?

**Comment by mikerenfro** [comment]
Resolved in 8b4908d ?

**Comment by gdevenyi** [comment]
Fixed by #1049 

Thanks!
